### PR TITLE
Remove failOnError from AssembleMojo

### LINKED ISSUE #2678: AssembleMojo.java:112-114: Remove the following...
The puzzle `2443-427f9946` from #2443 has to be resolved: 

https://github.com/objectionary/eo/blob/015f73814c9f158357866a07200b653cbcf0417d/eo-maven-plugin/src/main/java/org/eolang/maven/AssembleMojo.java#L112-L114

The puzzle was created by @rultor on 06-Dec-23. 

Estimate: 90 minutes,  role: DEV. 

If you have any technical questions, don't ask me, submit new tickets instead. The task will be \"done\" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

### PR DESCRIPTION
**Pull request by c71n93**
Closes #2678

---

## PR-Codex overview
This PR focuses on removing the `failOnError` flag in the `OptimizeMojo`, `VerifyMojo`, and `AssembleMojo` classes. 

### Detailed summary
- Removed the `failOnError` flag in the `OptimizeMojo` class.
- Removed the `failOnError` flag in the `VerifyMojo` class.
- Removed the `failOnError` flag in the `AssembleMojo` class.
- Replaced `Assertions.assertThrows` with `Assertions.assertDoesNotThrow` in the `SafeMojoTest` class.
- Updated the `OptimizeMojoTest` class to use `MatcherAssert.assertThat` instead of `Assertions.assertThrows`.
- Updated the `ParseMojoTest` class to use `MatcherAssert.assertThat` instead of `Assertions.assertThrows`.
- Updated the `VerifyMojoTest` class to use `Assertions.assertThrows` for testing failures.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by c71n93** [comment]
@maxonfjvipon could you check this one, please? Also do we need to `failOnError` from `ParseMojo` in this PR? Or it is better to create todo?

**Comment by maxonfjvipon** [comment]
@c71n93 the main goal of the ticket is to get rid of `failOnError` at all, in all mojos, so I think it would be better to do it in this PR

**Comment by c71n93** [comment]
@maxonfjvipon take a look, please

File: eo-maven-plugin/src/test/java/org/eolang/maven/AssembleMojoTest.java [hunk 1]
```diff
@@ -215,20 +215,12 @@ void assemblesTogetherWithVersions(@TempDir final Path temp) throws Exception {
...
-            result.get(String.format("target/%s", ParseMojo.DIR)),
-            new ContainsFiles(String.format("**/main.%s", TranspileMojo.EXT))
-        );
-        MatcherAssert.assertThat(
-            "Since the eo program invalid we shouldn't have optimized it, but we did",
-            result.get(String.format("target/%s", OptimizeMojo.DIR)),
-            Matchers.not(new ContainsFiles(String.format("**/main.%s", TranspileMojo.EXT)))
+    void failsOnInvalidProgram(@TempDir final Path temp) {
+        Assertions.assertThrows(
+            IllegalStateException.class,
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/AssembleMojoTest.java:220]: @c71n93 let's add a human readable assertion message here

**Review by maxonfjvipon — CHANGES_REQUESTED** [review]
@c71n93 good one, just one comment from my side

File: eo-maven-plugin/src/test/java/org/eolang/maven/AssembleMojoTest.java (hunk 1 above)
Comment (c71n93) [reply eo-maven-plugin/src/test/java/org/eolang/maven/AssembleMojoTest.java:220]: @maxonfjvipon done

**Review by maxonfjvipon — APPROVED** [review]
@c71n93 LGTM! Thanks

**Comment by c71n93** [comment]
@yegor256 could you check this one, please?

File: eo-maven-plugin/src/main/java/org/eolang/maven/ParseMojo.java [hunk 2]
```diff
@@ -190,18 +179,16 @@ private void parse(final ForeignTojo tojo) throws IOException {
...
             }
-            if (this.failOnError) {
-                throw new IllegalArgumentException(
-                    String.format(
-                        "Failed to parse %s (%d parsing errors)",
-                        source, errors.size()
-                    )
-                );
-            }
+            throw new IllegalArgumentException(
```
Comment (yegor256) [inline eo-maven-plugin/src/main/java/org/eolang/maven/ParseMojo.java:186]: @c71n93 I think, we should NOT fail here. We introduced the `verify` goal for this, right? We will fail there. If we fail here, the `verify` mojo will never be called. @maxonfjvipon am I right?

Comment (maxonfjvipon) [reply eo-maven-plugin/src/main/java/org/eolang/maven/ParseMojo.java:186]: @yegor256 you're right, I didn't notice that this code is in `ParseMojo`.
@c71n93 I think we need one more test, that parses with errors, but does not fail

**Review by maxonfjvipon — CHANGES_REQUESTED** [review]
@c71n93 see above

**Comment by c71n93** [comment]
@maxonfjvipon could you check again, please?

File: eo-maven-plugin/src/test/java/org/eolang/maven/ParseMojoTest.java [hunk 3]
```diff
@@ -117,29 +117,12 @@ void parsesWithCache(@TempDir final Path temp) throws Exception {
 
     @Test
     void crashesOnInvalidSyntax(@TempDir final Path temp) {
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/ParseMojoTest.java:119]: @c71n93 I think the test does not belong to the class. Here we should test `ParseMojo`, not `VerifyMojo`. Let's move it `VerifyMojoTest`. But here I believe we should add the test (if it does not exist yet) that would check that invalid program DOES NOT fail

File: eo-maven-plugin/src/test/java/org/eolang/maven/OptimizeMojoTest.java [hunk 4]
```diff
@@ -230,7 +230,7 @@ void failsOnCritical(@TempDir final Path temp) throws IOException {
                     "    TRUE > x",
                     "    FALSE > x"
                 ).with("trackOptimizationSteps", true)
-                .execute(new FakeMaven.Optimize())
+                .execute(new FakeMaven.Verify())
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/OptimizeMojoTest.java:233]: @c71n93 I think this whole test does not belong to the class. Here we should test `OptimizeMojo`, not `VerifyMojo`. Let's move the test to the `VerifyMojoTest`. But here I believe we should add the test (if it does not exist yet) that would check that program with errors DOES NOT fail on optimization step

File: eo-maven-plugin/src/test/java/org/eolang/maven/AssembleMojoTest.java [hunk 5]
```diff
@@ -215,32 +215,13 @@ void assemblesTogetherWithVersions(@TempDir final Path temp) throws Exception {
...
-        final Class<IllegalStateException> expected = IllegalStateException.class;
+    void failsOnInvalidProgram(@TempDir final Path temp) {
         Assertions.assertThrows(
-            expected,
+            IllegalStateException.class,
             () -> new FakeMaven(temp)
                 .withProgram(AssembleMojoTest.INVALID_PROGRAM)
-                .execute(AssembleMojo.class),
-            String.format("AssembleMojo should have failed with %s, but didn't", expected)
+                .execute(new FakeMaven.Verify()),
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/AssembleMojoTest.java:223]: I think this whole test does not belong to the class. Here we should test `AssembleMojo`, not `VerifyMojo`. Let's move the test to the `VerifyMojoTest`. But here I believe we should add the test (if it does not exist yet) that would check that program with errors DOES NOT fail on assemble step

**Review by maxonfjvipon — CHANGES_REQUESTED** [review]
@c71n93 see above

File: eo-maven-plugin/src/test/java/org/eolang/maven/ParseMojoTest.java (hunk 3 above)
Comment (c71n93) [reply eo-maven-plugin/src/test/java/org/eolang/maven/ParseMojoTest.java]: @maxonfjvipon done

File: eo-maven-plugin/src/test/java/org/eolang/maven/OptimizeMojoTest.java (hunk 4 above)
Comment (c71n93) [reply eo-maven-plugin/src/test/java/org/eolang/maven/OptimizeMojoTest.java]: @maxonfjvipon done

File: eo-maven-plugin/src/test/java/org/eolang/maven/AssembleMojoTest.java (hunk 5 above)
Comment (c71n93) [reply eo-maven-plugin/src/test/java/org/eolang/maven/AssembleMojoTest.java]: @maxonfjvipon done

**Review by maxonfjvipon — APPROVED** [review]
@c71n93 looks good to me

**Comment by c71n93** [comment]
@yegor256 could you check this one again, please?

**Comment by yegor256** [comment]
@rultor merge

**Pull request merged**